### ISSUE TITLE
Use squeeze during tpcc benchmark running may cause data inconsistent

### ISSUE THREAD

**Issue by jujing** [cybertec-postgresql/pg_squeeze#60]
If we use squeeze when we are running tpcc benchmark with benchmarksql, we found data inconsistent.
For exmaple, we squeeze table bmsql_order_line during tpcc benchmark running, the following sql should update ol_delivery_d feild of some records in table bmsql_order_line, but we found the ol_delivery_d values are still null after update:
UPDATE bmsql_order_line SET ol_delivery_d = $1 WHERE ((ol_w_id = $2) AND (ol_d_id = $3) AND (ol_o_id = $4)) 
parameters: $1 = '2023-04-27 19:02:29.481', $2 = '90', $3 = '1', $4 = '2175'

**Comment by jujing** [comment]
Is it the MVCC-unsafe behavior described in "Notes on Concurrency"?

**Comment by ahouska** [comment]
Yes, I could just verify (using debugger) that the conditions described in https://www.postgresql.org/docs/15/mvcc-caveats.html also apply to `UPDATE` and that `pg_squeeze` also causes this problem. I suppose you use the `REPEATABLE READ` transaction isolation. Doesn't it make sense to use `READ COMMITTED`?

To fix the problem on the extension side, `pg_squeeze` would have to perform "tuple rewriting". This isn't really trivial and there's currently no plan to introduce it.

**Comment by jujing** [comment]
I use READ COMMITTED transaction isolation. 
I guess the following conditions cause this problem:
1) Squeeze process got table AccessExclusiveLock lock to do final merge(tuple rewrite).
2) Update process got a snapshot then wait table release AccessExclusiveLock lock to get a lock for update.
3) Squeeze process finished final merge (tuple rewrite) then commit and release table AccessExclusiveLock lock.
4) Update process got the table lock for update, because the snapshot is got before squeeze process commit,  the tuples produced by squeeze process are not visible, so it update actually 0 recorders.

**Comment by jujing** [comment]
Use the following steps to reproduce this problem:
1. Use gdb attached to the squeeze process when executing "select squeeze.squeeze_table('public','bmsql_order_line',null,null)", break at swap_relation_files function.
2. Use another  process to execute update, likes: UPDATE bmsql_order_line SET ol_delivery_d = '2023-05-05 12:00:00.123' WHERE ((ol_w_id = 1) AND (ol_d_id = 5) AND (ol_o_id = 1127))
3. Continue the squeeze process to finish the table squeeze.
4. The update process finished update, and update 0 row. There are 8 rows should be updated but not.
5. Execute the same update sql again, it update 8 rows.

**Comment by ahouska** [comment]
If the update process acquired the `RowExclusiveLock` on the table (4) and if the transaction isolation is `READ COMMITTED`, it should get a new snapshot, which does see effects of `pg_squeeze`. I checked this even in PG 15 and PG 12. Which PG version are you using?

**Comment by jujing** [comment]
Yes, you are right, no problem when using  the REPEATABLE READ transaction isolation.
